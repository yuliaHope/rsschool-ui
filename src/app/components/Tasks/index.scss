.card,
.card-empty {
    width: 20vw;
    margin: 50px auto;
}

.card-title,
.card-empty-title {
    text-transform: uppercase;
}

.card-body {
    padding-bottom: 100px;
}

.score {
    float: right;
}

.submit-btn {
    background-color: #000;
    text-transform: uppercase;
    font-size: 14px;
    color: #fff;
}