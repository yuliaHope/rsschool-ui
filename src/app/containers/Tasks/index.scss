.tasks {
    padding: 20px;
}