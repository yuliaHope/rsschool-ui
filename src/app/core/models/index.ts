export * from './events';
export * from './stage';
export * from './user';
export * from './course';
export * from './assignments';
export * from './courseMentor';
export * from './courseStudent';
