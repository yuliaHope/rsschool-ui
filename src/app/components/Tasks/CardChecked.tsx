import * as React from 'react';
import { classNames } from 'core/styles';
import { Card, CardHeader, CardBody, CardTitle, CardText } from 'reactstrap';

const cn = classNames(require('./index.scss'));

type Props = {
    score: number;
    title: string;
    description: string;
};

const CardChecked = (props: Props) => {
    const { score, title, description } = props;
    return (
        <Card className={`${cn('card')} ${score >= 100 ? cn('border-success') : cn('border-warning')}`}>
            <CardHeader className={`${cn('text-white')} ${score >= 100 ? cn('bg-success') : cn('bg-warning')}`}>
                {score >= 100 ? 'Done!' : 'Not submitted yet!'} <span className={cn('score')}>{score}%</span>
            </CardHeader>
            <CardBody className={cn('card-body')}>
                <CardTitle className={cn('card-title')}>{title}</CardTitle>
                <CardText>{description}</CardText>
            </CardBody>
        </Card>
    );
};

export default CardChecked;
