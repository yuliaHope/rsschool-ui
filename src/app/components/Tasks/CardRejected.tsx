import * as React from 'react';
import { classNames } from 'core/styles';
import { Card, CardHeader, CardBody, CardTitle, CardText } from 'reactstrap';

const cn = classNames(require('./index.scss'));

type Props = {
    title: string;
    description: string;
};

const CardRejected = (props: Props) => {
    const { title, description } = props;
    return (
        <Card className={`${cn('card')} ${cn('border-danger')}`}>
            <CardHeader className={cn('text-white bg-danger')}>The deadline has passed</CardHeader>
            <CardBody className={cn('card-body')}>
                <CardTitle className={cn('card-title')}>{title}</CardTitle>
                <CardText>{description}</CardText>
            </CardBody>
        </Card>
    );
};

export default CardRejected;
