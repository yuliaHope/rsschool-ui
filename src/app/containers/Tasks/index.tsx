import * as React from 'react';
import { TaskItem } from 'components/TaskItem/TaskItem';
import { connect } from 'react-redux';
import { RootState } from 'core/reducers';
import { fetchAssignments, fetchSolution } from 'core/actions';
import { IAssignmentModel } from 'core/models';
type Props = {
    courseId: string;
    userGitId: string;
    assignments?: Array<IAssignmentModel>;
    loading: boolean;
    error: any;
    solutionError: any;
    fetchAssignmentsData: (courseId: string) => void;
    fetchAssignmentSolution: (assignment: IAssignmentModel) => void;
};

const mapStateToProps = (state: RootState, props: any): Props => {
    if (state.assignments == null || state.user == null) {
        return props;
    }
    return {
        ...props,
        userGitId: state.user.username,
        courseId: props.match.params.id,
        assignments: state.assignments.assignments,
        loading: state.assignments.loading,
        error: state.assignments.error,
        solutionError: state.assignments.solutionError,
    };
};

const mapDispatchToProps = (dispatch: any, props: any): Props => {
    return {
        ...props,
        fetchAssignmentsData: (courseId: string) => {
            dispatch(fetchAssignments(courseId));
        },
        fetchAssignmentSolution: (assignment: IAssignmentModel) => {
            dispatch(fetchSolution(assignment));
        },
    };
};

class Tasks extends React.Component<Props> {
    componentDidMount() {
        this.props.fetchAssignmentsData(this.props.courseId);
    }
    getFullScore(): number {
        const { assignments } = this.props;
        let score = 0;
        if (assignments && assignments.length !== 0) {
            assignments.map(elem => {
                score += elem.score;
            });
        }
        return score;
    }
    render() {
        const { assignments, error, fetchAssignmentSolution, userGitId } = this.props;
        return (
            <React.Fragment>
                <h2>Tasks</h2>
                <div className="row">
                    <div className="col-6">
                        <p>
                            Your github private repository
                            <a className="badge badge-dark" href={`https://github.com/${userGitId}`}>
                                Link TO Git
                            </a>
                        </p>
                    </div>
                    <div className="col-6 text-right">
                        <p>You are in the TOP 50 students!</p>
                        <p>Full Score:{this.getFullScore()}</p>
                    </div>
                </div>
                <div className="card-deck mb-3">
                    {assignments ? (
                        assignments.map(elem => (
                            <TaskItem
                                assignment={elem}
                                key={elem.taskId}
                                fetchAssignmentSolution={fetchAssignmentSolution}
                            />
                        ))
                    ) : (
                        <div>Loading</div>
                    )}
                    {error ? <div>Error</div> : null}
                </div>
            </React.Fragment>
        );
    }
}

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(Tasks);
