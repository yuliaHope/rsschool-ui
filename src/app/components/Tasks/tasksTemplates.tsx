import { IAssignment, IAssignmentStatus } from 'core/models';
import * as React from 'react';
import CardChecked from './CardChecked';
import CardRejected from './CardRejected';

export enum FeedActions {
    'Assignment' = 'Assignment',
}

export const tasksTemplates: any = {
    [FeedActions.Assignment]: {
        [IAssignmentStatus.Checked]: (assignment: IAssignment): any => {
            return (
                <CardChecked
                    score={assignment.score}
                    title={assignment.taskId}
                    description={assignment.assignmentRepo}
                />
            );
        },
        [IAssignmentStatus.Rejected]: (assignment: IAssignment): any => {
            return <CardRejected title={assignment.taskId} description={assignment.assignmentRepo} />;
        },
    },
};
