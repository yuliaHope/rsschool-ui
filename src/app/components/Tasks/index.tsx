import * as React from 'react';
import { IAssignment } from 'core/models';
import { tasksTemplates } from './tasksTemplates';
import { classNames } from 'core/styles';
const cn = classNames(require('./index.scss'));
import {
    Card,
    CardHeader,
    CardBody,
    CardFooter,
    CardTitle,
    CardText,
    Form,
    FormGroup,
    Label,
    Input,
    Button,
} from 'reactstrap';

type Props = {
    assign: () => void;
    data: IAssignment[];
};

class Tasks extends React.PureComponent<Props> {
    hasData() {
        return Array.isArray(this.props.data) && this.props.data.length > 0;
    }

    render() {
        return (
            <div>
                <h2>Tasks</h2>
                <div className="row">
                    <div className="col-6">
                        <p>
                            Your github private repository{' '}
                            <a className="badge badge-dark" href="#">
                                here
                            </a>
                        </p>
                    </div>
                    <div className="col-6 text-right">
                        <p>You are in the TOP X students!</p>
                        <p>Full Score: X</p>
                    </div>
                </div>

                {this.hasData() ? (
                    this.props.data.map((assignment, i) => {
                        const template = tasksTemplates[assignment.taskId][assignment.status];
                        return (
                            <div key={i}>
                                <div>
                                    {template ? (
                                        template(assignment)
                                    ) : (
                                        <Card className={cn('card-empty')}>
                                            <CardHeader>Not submitted yet!</CardHeader>
                                            <CardBody>
                                                <CardTitle className={cn('card-empty-title')}>
                                                    {this.props.data[i].taskId}
                                                </CardTitle>
                                                <CardText>{this.props.data[i].assignmentRepo}</CardText>
                                            </CardBody>
                                            <CardFooter>
                                                <small className={cn('text-muted')}>
                                                    <Form>
                                                        <FormGroup>
                                                            <Label for="taskUrl">Choose repo</Label>
                                                            <Input
                                                                bsSize="sm"
                                                                id="taskUrl"
                                                                name="url"
                                                                placeholder="Enter link"
                                                                required={true}
                                                                type="url"
                                                            />
                                                        </FormGroup>
                                                        <FormGroup>
                                                            <Label for="taskText">Comments</Label>
                                                            <Input
                                                                bsSize="sm"
                                                                id="taskText"
                                                                name="text"
                                                                placeholder="Write comments here"
                                                                rows="5"
                                                                type="textarea"
                                                            />
                                                        </FormGroup>
                                                        <Button
                                                            onClick={this.props.assign}
                                                            className={cn('submit-btn')}
                                                        >
                                                            Submit
                                                        </Button>
                                                    </Form>
                                                </small>
                                            </CardFooter>
                                        </Card>
                                    )}
                                </div>
                            </div>
                        );
                    })
                ) : (
                    <div>Nothing to show</div>
                )}
            </div>
        );
    }
}

export default Tasks;
