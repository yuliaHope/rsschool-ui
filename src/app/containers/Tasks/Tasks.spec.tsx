import { shallow } from 'enzyme';
import * as React from 'react';
import Tasks from './index';
import { AssignmentStatus } from '../../core/models';
const { __stateGetter } = require('react-redux');

describe('Tasks container', () => {
    beforeEach(() => {
        __stateGetter.mockImplementation(() => ({}));
    });
    const props = {
        courseId: 'courseId',
        userGitId: 'userGitId',
        assignments: [],
        loading: false,
        error: null,
        solutionError: null,
        fetchAssignmentsData: jest.fn(),
        fetchAssignmentSolution: jest.fn(),
    };
    it('render tasks container', () => {
        const output = shallow(<Tasks {...props} />).dive();
        expect(output).toMatchSnapshot();
    });

    it('dispatches fetchAssignmentsData', () => {
        expect(props.fetchAssignmentsData).toHaveBeenCalledTimes(1);
    });
    it('render with preloader', () => {
        const nextProps = {
            ...props,
            loading: true,
        };
        const output = shallow(<Tasks {...nextProps} />).dive();
        expect(output).toMatchSnapshot();
    });
    it('render properly', () => {
        const nextProps = {
            ...props,
            loading: false,
            assignments: [
                {
                    _id: '_id',
                    assignmentRepo: 'assignmentRepo',
                    checkDate: 123,
                    completeDate: 123,
                    courseId: 'courseId',
                    deadlineDate: 123,
                    mentorComment: 'mentorComment',
                    mentorId: 'mentorId',
                    score: 90,
                    status: AssignmentStatus.Assigned,
                    studentComment: 'studentComment',
                    studentId: 'studentId',
                    taskId: 'taskId',
                    title: 'title',
                },
            ],
        };
        const output = shallow(<Tasks {...nextProps} />).dive();
        expect(output).toMatchSnapshot();
    });
});
