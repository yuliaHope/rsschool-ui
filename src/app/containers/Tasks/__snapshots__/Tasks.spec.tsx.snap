// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks container render properly 1`] = `
<React.Fragment>
  <h2>
    Tasks
  </h2>
  <div
    className="row"
  >
    <div
      className="col-6"
    >
      <p>
        Your github private repository
        <a
          className="badge badge-dark"
          href="https://github.com/userGitId"
        >
          Link TO Git
        </a>
      </p>
    </div>
    <div
      className="col-6 text-right"
    >
      <p>
        You are in the TOP 50 students!
      </p>
      <p>
        Full Score:
        90
      </p>
    </div>
  </div>
  <div
    className="card-deck mb-3"
  >
    <Component
      assignment={
        Object {
          "_id": "_id",
          "assignmentRepo": "assignmentRepo",
          "checkDate": 123,
          "completeDate": 123,
          "courseId": "courseId",
          "deadlineDate": 123,
          "mentorComment": "mentorComment",
          "mentorId": "mentorId",
          "score": 90,
          "status": "Assigned",
          "studentComment": "studentComment",
          "studentId": "studentId",
          "taskId": "taskId",
          "title": "title",
        }
      }
      fetchAssignmentSolution={[MockFunction]}
      key="taskId"
    />
  </div>
</React.Fragment>
`;

exports[`Tasks container render tasks container 1`] = `
<React.Fragment>
  <h2>
    Tasks
  </h2>
  <div
    className="row"
  >
    <div
      className="col-6"
    >
      <p>
        Your github private repository
        <a
          className="badge badge-dark"
          href="https://github.com/userGitId"
        >
          Link TO Git
        </a>
      </p>
    </div>
    <div
      className="col-6 text-right"
    >
      <p>
        You are in the TOP 50 students!
      </p>
      <p>
        Full Score:
        0
      </p>
    </div>
  </div>
  <div
    className="card-deck mb-3"
  />
</React.Fragment>
`;

exports[`Tasks container render with preloader 1`] = `
<React.Fragment>
  <h2>
    Tasks
  </h2>
  <div
    className="row"
  >
    <div
      className="col-6"
    >
      <p>
        Your github private repository
        <a
          className="badge badge-dark"
          href="https://github.com/userGitId"
        >
          Link TO Git
        </a>
      </p>
    </div>
    <div
      className="col-6 text-right"
    >
      <p>
        You are in the TOP 50 students!
      </p>
      <p>
        Full Score:
        0
      </p>
    </div>
  </div>
  <div
    className="card-deck mb-3"
  />
</React.Fragment>
`;
